Cirrus: Add netavark/aardvark system test task

cevich opened the pull request:
***Depends on:*** #12814

Luap99 commented:
I think you have to remove the `99-do-not-use-google-subnets.conflist` file, otherwise the network backend detection logic thinks that you want to use CNI because you already have extra CNI networks.

cevich commented:
> already have extra CNI networks.

Oh gosh @Luap99 thank you so much.  I've been struggling to figure this out.  Oddly enough, when I go in and run the test manually, it works (after running the exact same `setup_environment.sh`).  I'll try what you suggest though, it's a good idea...

edsantiago commented:
Still all-red CI, so I'm not likely to be able to review this before the weekend. But should you be interested, here's the new task map:
![cirrus-map-pr13024](https://user-images.githubusercontent.com/265639/151441357-1a47dfca-9f01-43a1-ba39-5cb34af795ec.png)
The new task is the pukey-lime one near six o'clock. I'm curious about its location: shouldn't it come after integration tests?

cevich commented:
>  I'm not likely to be able to review

I think that's okay, I'm questioning if I can even get it working.  It's working fine under `hack/get_ci_vm.sh` but not normally...very strange since the environments are 99.999% the same (by design).

> shouldn't it come after integration tests?

The task map is going to be all f'd up while me `DO NOT MERGE` commit is in place.  It's messing with things to get the new task running as soon/quickly as possible.

cevich commented:
Okay, the basic `podman info` system test is working now (thanks @Luap99 for spotting the problem!).  This shows we've got some (likely) netavark-related breakage in other tests:

https://cirrus-ci.com/task/5138206667571200?logs=main#L439

@baude @edsantiago thoughts on skipping them for now (I'll file new issues) or should we take some other PR strategy?

edsantiago commented:
@cevich please make sure there are issues filed for the problems, then:
```bash
    if running-under-netavark; then
      skip "FIXME: this test doesn't work under nevavark; #12345"
   fi
```
(or, if just skipping a portion of a test, include FIXME and #12345 on the same line in a comment. That makes it easy for my `find-obsolete-skips` script to alert me about these when #12345 gets fixed)

Luap99 commented:
First test aardvark doesn't seem to work, we have to investigate why.
The second one is expected because why need to change the iptables command to flush netavark rules.

cevich commented:
Great, thanks @Luap99 and @edsantiago .  Yep, I've already got an `is_netavark()` function ready.  I'll make sure to open issues on anything I mark `skip()` with it (seems like just two things - which is pretty good, all things considered).

cevich commented:
[Clean run after skipping the two failing tests (I opened issues for them)](https://cirrus-ci.com/task/5179305075408896).  Rebased, squashed, and force-pushed w/o the `DO NOT MERGE` commit.  Fingers crossed :crossed_fingers:

Luap99 commented on test/system/helpers.bash:
```diff
@@ -341,6 +341,13 @@ function is_cgroupsv2() {
+function is_netavark() {
+    # FIXME: Do the tests ever run against other versions?
+    # Can't safely use --filter: < v4.0 do not have the networkBackend field.
+    $PODMAN info | grep -qiw 'networkBackend:.*netavark'
```
Why do you not use `podman info --format {{.Host.NetworkBackend}}`. The test always have to match the podman version. There is no point in running the current system tests with an older version as they will obviously fail because new tests have been added.

cevich replied:
Okay, right.  Ya I wasn't sure about that.  Can certainly use `--format` then.

cevich replied:
Okay, this is in place now.

Luap99 commented on test/system/500-networking.bats:
```diff
@@ -224,6 +227,9 @@ load helpers
 @test "podman network reload" {
+    if is_netavark; then
+        skip "Broken with netavark: https://github.com/containers/podman/issues/13073"
+    fi
```
see https://github.com/containers/podman/issues/13073#issuecomment-1026021580

cevich replied:
Thanks for the help.  I'm going to step away from this PR for a bit (since the int. tests are failing).

cevich commented:
Giving @Luap99 network-reload test fix suggestion a try.

cevich commented:
[Example of netavark/aardvark build details.](https://cirrus-ci.com/task/5944241670062080?logs=package_versions#L16-L29)

Luap99 commented on test/system/500-networking.bats:
```diff
@@ -163,6 +163,7 @@ load helpers
 
 # "network create" now works rootless, with the help of a special container
 @test "podman network create" {
+    skip_if_netavark "Broken with netavark: https://github.com/containers/podman/issues/13072"
```
Can remove this skip and try again I think we fixed many aardvark bugs and this works locally for me.

cevich replied:
Happy to, but I think I need to wait for https://github.com/containers/podman/pull/12859 to merge, since (I'm assuming) it also fixes the `Netavark integaration` failures happening here.

cevich replied:
okay, I removed the commit that added this, lets see how it goes...

edsantiago commented:
Shouldn't the new task include `CI:BUILD`? Also, would it make sense to drag it up and run after `netavark_integration_test` instead of `local_integration_test`? Just would seem to look cleaner on the task map that way.
![cirrus-map-pr13024](https://user-images.githubusercontent.com/265639/152243446-0dabf98a-4b57-4507-b091-0747ba5b9e47.png)

cevich commented:
Oh, good catches @edsantiago fixing that now...

edsantiago commented:
It's really obvious in the diagram :-)

cevich commented:
Done.  I also re-arranged the dependency list in `success` to match the task definition order.

edsantiago commented:
That looks better, thanks. New map shows it in fuchsia near top right. 
![cirrus-map-pr13024](https://user-images.githubusercontent.com/265639/152244947-d2f318ec-3fe8-4a20-8e92-4ef0c7bb13bb.png)

edsantiago commented on test/system/helpers.bash:
```diff
@@ -341,6 +341,16 @@ function is_cgroupsv2() {
+function is_netavark() {
+    local be_type
+    be_type=$($PODMAN info --format '{{.Host.NetworkBackend}}')
+    if [[ "$be_type" =~ netavark ]]; then
+        return 0
+    fi
+    return 1
```
Could be shortened and legibilitized as just
```
test "$be_type" =~ "netavark"
```

cevich replied:
I'll change it so it's consistent with the style of other functions.  However, my understanding is that `[[` is a bash built-in,  therefor likely MUCH faster than executing the `test` binary.  I'd also argue that keeping things less readable in this instance is actually an asset. Since it keeps the bash-charlatans and bash-wannabe's far away :rofl: Lol, bash, legible, funny stuff :rofl:

edsantiago commented on test/system/helpers.bash:
```diff
@@ -341,6 +341,16 @@ function is_cgroupsv2() {
+# True if podman is using netavark
+function is_netavark() {
+    local be_type
+    be_type=$($PODMAN info --format '{{.Host.NetworkBackend}}')
```
Is the use of `$PODMAN` here deliberate? (As opposed to `run_podman` and using `$output`)? The only important reason to use `$PODMAN` is when it's ultra-important to use the local client version instead of `-remote`, and I don't think in this case it makes a difference. This whole function could be two lines:
```
run_podman info --format ...
test "$output" =~ "netavark"
```

cevich replied:
Yes it was on purpose.  I didn't want any chance of secondary issues conflating the (simple) query with complexities of a complex function.  It's also much faster this way, and seemed to be the pattern of all the other `is_XYZ()` functions, so I went with it.  I'll change it, but it's only out of expediency getting this PR merged.

edsantiago commented on test/system/005-info.bats:
```diff
@@ -88,6 +88,18 @@ host.slirp4netns.executable | $expr_path
 }
 
+@test "podman info netavark " {
+    # Confirm podman uses netavark only when requested
```
Careful with this "only", it is misleading.
* If you really truly mean "only", then you should do the converse test: in the `else` clause, confirm that `is_netavark` is false.
* Otherwise, please remove the `only`. That makes it clearer that it's OK to use netavark even if `NETWORK_BACKEND` is not set or if it's set to something else.

edsantiago commented on test/system/helpers.bash:
```diff
@@ -473,6 +483,15 @@ function skip_if_root_ubuntu {
     fi
 }
 
+######################
+#  skip_if_netavark  #  ...with an optional message
+######################
+function skip_if_netavark() {
+    if is_netavark; then
+        skip "${1:-not supported with netavark network backend}"
+    fi
+}
+
```
This is not currently used. Could we just delete it, and assume that someone will add it if the need arises?

cevich replied:
sure.

edsantiago reviewed (commented):
Partial LGTM, with a few requests

edsantiago commented:
Looks like I should've tested my suggestion, sorry:
```
# helpers.bash: line 347: test: =~: binary operator expected
```

cevich commented:
Mass-re-run failed tasks: Something broke with quay.

edsantiago commented:
I've been fighting what I think are CDN failures all day. Maybe they're related.

cevich commented:
> CDN failures

IKR.  It's like the *ONE THING* we have CDNs for, and they're somehow not redundant enough (or some network broke)  :cry:

cevich commented:
Just looked in Cirrus WebUI and things are progressing now (despite what the github UI says).

cevich commented:
Yay :partying_face: quick merge it before the universe changes it's mind :smile:

edsantiago commented:
/lgtm

Luap99 commented:
/lgtm

rhatdan commented:
/approve
